File: preview/systems/preview.cpp
// ============================================================
//   ARQUIVO DE TESTE - PREVIEW DO TEMA "Esron Dark"
//   Cada bloco comentado indica a categoria semântica testada
// ============================================================

#include <cstdio>
#include <cstdlib>
#include <string>
#include <regex>

// ---------- Macros ----------
#define MAX_TENTATIVAS 5
#define QUADRADO(x) ((x) * (x))

// ---------- Namespace ----------
namespace app
{
    // ---------- Tipos primitivos (esperado: roxo, sem negrito) ----------
    int contador = 0;
    char letra = 'A';
    double preco = 19.90;
    float media = 7.5f;
    bool ativo = true;
    unsigned int codigo = 42;
    long total = 1000000L;
    void* ponteiro = nullptr;

    // ---------- Struct (esperado: tipo em vermelho suave D16D6D) ----------
    struct Ponto
    {
        int x;   // "int" roxo | "x" propriedade
        int y;
    };

    // ---------- Enum simples ----------
    enum Cor
    {
        VERMELHO,
        VERDE,
        AZUL
    };

    // ---------- Enum class ----------
    enum class Status
    {
        Ativo,
        Inativo,
        Pendente
    };

    // ---------- Classe + herança ----------
    class Pessoa
    {
        private:
            std::string nome;
            int idade;

        public:
            Pessoa(std::string n, int i) : nome(n), idade(i) {}

            std::string getNome() const
            {
                return nome;
            }

            int getIdade() const
            {
                return idade;
            }
    };

    class Funcionario : public Pessoa
    {
        public:
            double salario;

            Funcionario(std::string n, int i, double s)
                : Pessoa(n, i), salario(s) {}
    };
}

// ---------- Variável global (esperado: vermelho vivo EF5350) ----------
int variavelGlobal = 100;

// ---------- Constante / readonly (esperado: vermelho mais fechado BF616A) ----------
const int LIMITE_MAXIMO = 999;
constexpr int VALOR_CONSTANTE = 42;

// ---------- Variável estática (testa modifier *.static) ----------
static int contadorEstatico = 0;

// ---------- Função do usuário (esperado: azul #5A7FBF) ----------
int somar(int a, int b)          // parâmetros a, b
{
    int resultado = a + b;       // variável local, vermelho vivo
    return resultado;
}

double calcularMedia(double valores[], int tamanho)
{
    double soma = 0;
    for (int i = 0; i < tamanho; i++)
    {
        soma += valores[i];
    }
    return soma / tamanho;
}

int main()
{
    // ---------- Funções da biblioteca padrão (esperado: ciano #88C0D0) ----------
    printf("Hello world!\n");
    system("pause");

    // ---------- Strings ----------
    std::string mensagem = "Testando o tema Esron Dark";

    // ---------- Números ----------
    int numero = 42;
    double pi = 3.14159;
    int hexadecimal = 0xFF;

    // ---------- Booleanos ----------
    bool sucesso = true;
    bool falha = false;

    // ---------- Operadores ----------
    int soma = numero + 10;
    bool comparacao = (soma > 0) && (falha == false);

    // ---------- Enum member em uso ----------
    app::Cor corEscolhida = app::VERMELHO;
    app::Status statusAtual = app::Status::Ativo;

    // ---------- Struct em uso ----------
    app::Ponto p1;
    p1.x = 10;
    p1.y = 20;

    // ---------- Classe em uso ----------
    app::Pessoa pessoa("Esron", 30);
    printf("Nome: %s\n", pessoa.getNome().c_str());

    // ---------- Macro em uso ----------
    int quadrado = QUADRADO(5);

    // ---------- Regex ----------
    std::regex padrao("[0-9]+");
    bool contemNumero = std::regex_search(mensagem, padrao);

    // ---------- Condicional ----------
    if (soma > 0)
    {
        printf("soma positiva\n");
    }
    else
    {
        printf("soma negativa\n");
    }

    // ---------- Laço ----------
    for (int i = 0; i < MAX_TENTATIVAS; i++)
    {
        printf("tentativa %d\n", i);
    }

    /* Comentário
       de múltiplas
       linhas */

    return 0;
}

File: preview/web/preview.js
// ============================================================
//   ARQUIVO DE TESTE - PREVIEW DO TEMA "Esron Dark"
//   Cada bloco comentado indica a categoria testada
// ============================================================

// ---------- Import (esperado: keyword) ----------
import { readFile } from "fs";

// ---------- var / let / const ----------
var legado = "não use var";
let contador = 0;
const LIMITE_MAXIMO = 999; // esperado: vermelho fechado (readonly)

// ---------- "Enum" simulado (JS não tem enum nativo) ----------
const Cor = Object.freeze({
    VERMELHO: "vermelho",
    VERDE: "verde",
    AZUL: "azul",
});

// ---------- Classe (esperado: tipo em vermelho suave D16D6D) ----------
class Pessoa {
    #idadePrivada; // campo privado

    static contadorDeInstancias = 0; // campo static

    constructor(nome, idade) {
        this.nome = nome;
        this.#idadePrivada = idade;
        Pessoa.contadorDeInstancias++;
    }

    get idade() {
        return this.#idadePrivada;
    }

    set idade(valor) {
        this.#idadePrivada = valor;
    }

    apresentar() {
        return `${this.nome}, ${this.idade} anos`;
    }

    static criarAnonimo() {
        return new Pessoa("Anônimo", 0);
    }
}

// ---------- Herança ----------
class Funcionario extends Pessoa {
    constructor(nome, idade, salario) {
        super(nome, idade);
        this.salario = salario;
    }

    apresentar() {
        return `${super.apresentar()} - R$ ${this.salario}`;
    }
}

// ---------- Função declarada (esperado: azul #5A7FBF) ----------
function somar(a, b) {
    // parâmetros a, b (branco)
    const resultado = a + b; // variável local (vermelho vivo)
    return resultado;
}

// ---------- Função expressão + arrow function ----------
const multiplicar = function (a, b) {
    return a * b;
};

const dividir = (a, b) => a / b;

// ---------- Função geradora ----------
function* contarAte(max) {
    for (let i = 0; i < max; i++) {
        yield i;
    }
}

// ---------- Função assíncrona (testa modifier async) ----------
async function calcularAsync(valor) {
    await new Promise((resolve) => setTimeout(resolve, 100));
    return valor * 2;
}

// ---------- Closures ----------
function criarContador() {
    let interno = 0;
    return function () {
        return ++interno;
    };
}

async function main() {
    // ---------- Funções de biblioteca (esperado: ciano #88C0D0) ----------
    console.log("Hello world!");
    const raiz = Math.sqrt(16);

    // ---------- Strings ----------
    const mensagem = "Testando o tema Esron Dark";
    const interpolada = `Mensagem: ${mensagem}`;

    // ---------- Números ----------
    const numero = 42;
    const pi = 3.14159;
    const hexadecimal = 0xff;
    const grande = 9007199254740991n; // BigInt

    // ---------- Booleanos / null / undefined ----------
    const sucesso = true;
    const falha = false;
    const vazio = null;
    let indefinido;

    // ---------- Operadores ----------
    const soma = numero + 10;
    const comparacao = soma > 0 && falha === false;

    // ---------- Optional chaining / nullish coalescing ----------
    const objeto = { config: { tema: "dark" } };
    const tema = objeto?.config?.tema ?? "claro";

    // ---------- Destructuring ----------
    const { nome, idade } = { nome: "Esron", idade: 30 };
    const [primeiro, segundo, ...resto] = [1, 2, 3, 4, 5];

    // ---------- Spread ----------
    const listaOriginal = [1, 2, 3];
    const listaCopia = [...listaOriginal, 4, 5];

    // ---------- Enum simulado em uso ----------
    const corEscolhida = Cor.VERMELHO;

    // ---------- Classes em uso ----------
    const pessoa = new Funcionario("Esron", 30, 5000.0);
    console.log(pessoa.apresentar());

    // ---------- Array methods (builtin) ----------
    const numeros = [1, 2, 3, 4, 5];
    const pares = numeros.filter((n) => n % 2 === 0).map((n) => n * 2);

    // ---------- Map / Set ----------
    const mapa = new Map();
    mapa.set("chave", "valor");

    const conjunto = new Set([1, 2, 2, 3]);

    // ---------- Regex ----------
    const padrao = /[0-9]+/;
    const contemNumero = padrao.test(mensagem);

    // ---------- JSON ----------
    const json = JSON.stringify({ nome, idade });

    // ---------- Generator em uso ----------
    for (const valor of contarAte(3)) {
        console.log(valor);
    }

    // ---------- Async em uso ----------
    const resultadoAsync = await calcularAsync(21);

    // ---------- Condicional ----------
    if (soma > 0) {
        console.log("soma positiva");
    } else {
        console.log("soma negativa");
    }

    // ---------- Laço ----------
    for (let i = 0; i < 5; i++) {
        console.log(`tentativa ${i}`);
    }

    // ---------- Exceção ----------
    try {
        JSON.parse("{ inválido");
    } catch (erro) {
        console.log("erro:", erro.message);
    }

    /* Comentário
       de múltiplas
       linhas */
}

main();

// ---------- Export ----------
export { somar, Pessoa };